### PR TITLE
Issue 2786 upload terminology command add explicit current version control

### PR DESCRIPTION
**Pull request by jmarchionatto**

### CODE REVIEW

**Review by IanMMarshall — APPROVED** [review]
Looks good. A few minor suggestions but no concerns. Approved.

File: hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/UploadTerminologyCommand.java [hunk 1]
```diff
@@ -132,6 +140,16 @@ private void invokeOperation(String theTermUrl, String[] theDatafile, IGenericCl
 
 		ParametersUtil.addParameterToParametersUri(myFhirCtx, inputParameters, TerminologyUploaderProvider.PARAM_SYSTEM, theTermUrl);
 
+		if (StringUtils.isNotBlank(makeCurrentVersionOption)) {
+			boolean isMakeCurrentVersion = Boolean.parseBoolean(makeCurrentVersionOption);
+			ParametersUtil.addParameterToParametersString(
+				myFhirCtx, inputParameters, TermLoaderSvcImpl.MAKE_CURRENT_VERSION, String.valueOf(isMakeCurrentVersion));
+		}
+		//todo: is this the case?. Check with client
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/UploadTerminologyCommand.java:148]: That as intention, yes. I just didn't want to delay review until meeting with client.

File: hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/UploadTerminologyCommand.java [hunk 2]
```diff
@@ -76,6 +79,8 @@ public Options getOptions() {
 		addRequiredOption(options, "u", "url", true, "The code system URL associated with this upload (e.g. " + ITermLoaderSvc.SCT_URI + ")");
 		addOptionalOption(options, "d", "data", true, "Local file to use to upload (can be a raw file or a ZIP containing the raw file)");
 		addOptionalOption(options, "m", "mode", true, "The upload mode: SNAPSHOT (default), ADD, REMOVE");
+		addOptionalOption(options, null, CURRENT_VERSION, true, "For LOINC upload. If specified as 'false' the uploaded version will not become current. " +
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/UploadTerminologyCommand.java:82]: I worded it this way because the default value is 'true', however, Frank Tao pointed me that as this flag is only for LOINC processing it would be better to have it as a property in the loinc.properties file, which I think is a great suggestion, and will eliminate the parameter for the utility.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/TerminologyUploaderProvider.java [hunk 3]
```diff
@@ -102,6 +102,7 @@ public TerminologyUploaderProvider(FhirContext theContext, ITermLoaderSvc theTer
 	public IBaseParameters uploadSnapshot(
 		HttpServletRequest theServletRequest,
 		@OperationParam(name = PARAM_SYSTEM, min = 1, typeName = "uri") IPrimitiveType<String> theCodeSystemUrl,
+		@OperationParam(name = TermLoaderSvcImpl.MAKE_CURRENT_VERSION, min = 0, max = 1) IPrimitiveType<String> isMakeCurrentVersion,
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/provider/TerminologyUploaderProvider.java:105]: It is used later at the called methods, however it will be removed when updating code to use property instead of parameter.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermLoaderSvcImpl.java [hunk 4]
```diff
@@ -245,31 +259,57 @@ public UploadStatistics loadLoinc(List<FileDescriptor> theFiles, RequestDetails
...
+
+		List<String> values = ParametersUtil.getNamedParameterValuesAsString(FhirContext.forR4(), parameters, MAKE_CURRENT_VERSION);
+		if (CollectionUtils.isEmpty(values))  return true;
+
+		if (values.size() > 1) {
+			throw new InvalidRequestException("Parameter '" + MAKE_CURRENT_VERSION + "' was specified more than once.");
+		}
+
+		if (! values.get(0).equalsIgnoreCase("false") && ! values.get(0).equalsIgnoreCase("true")) {
+			throw new InvalidRequestException("Parameter " + MAKE_CURRENT_VERSION + " only accepts values: ['true' or 'false']");
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermLoaderSvcImpl.java:303]: I think this code will be gone when changing parameter by property. If that is not the case, will add unit test

File: hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/UploadTerminologyCommand.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (IanMMarshall) [reply hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/UploadTerminologyCommand.java:82]: OK, good enough. I made the suggestion because when I first read it, the double-negative seemed a bit awkward. How about this wording instead:

> If specified as 'false' the uploaded version will not replace the current version.

In regards to specifying this as a flag in the loinc.properties, I agree entirely with this approach.

[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-cli/hapi-fhir-cli-api/src/main/java/ca/uhn/fhir/cli/UploadTerminologyCommand.java:82]: Great. Will do that improvement, then.

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/TerminologyLoaderSvcLoincTest.java [hunk 5]
```diff
@@ -822,6 +837,109 @@ public void testLoadLoincMultiaxialHierarchySupport() throws Exception {
 		assertEquals("LP52960-9", doublyNestedChildCode.getChildren().get(2).getChild().getCode());
 	}
 
+
+
+
+	@Nested
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/TerminologyLoaderSvcLoincTest.java:843]: Great feature IMO. https://www.youtube.com/watch?v=VaDAlqwCZIg
